examples/ecs/observers.rs

```diff
@@ -0,0 +1,198 @@
+    world.init_component::<TriggerMines>();
+    world.init_component::<Explode>();
+
+    // Observers are triggered when a certain event it fired, each event is represented by a component type.
+    // This observer runs whenever `TriggerMines` is fired, observers run systems which can be defined as closures.
+    world.observer(
+        |observer: Observer<TriggerMines>,
+         mines: Query<&Mine>,
+         index: Res<SpatialIndex>,
+         mut commands: Commands| {
+            // You can access the event data via the `Observer`
+            let trigger = observer.data();
+            // Access resources
+            for e in index.get_nearby(trigger.pos) {
+                // Run queries
+                let mine = mines.get(e).unwrap();
+                if mine.pos.distance(trigger.pos) < mine.size + trigger.radius {
+                    // And queue commands, including firing additional events
+                    // Here we fire the `Explode` event at entity `e`
+                    commands.event(Explode).entity(e).emit();
+                }
+            }
+        },
+    );
+
+    // Observers can also listen for events triggering for a specific component.
+    // This observer runs whenever the `Mine` component is added to an entity, and places it in a simple spatial index.
+    world.observer(
+        |observer: Observer<OnAdd, Mine>, query: Query<&Mine>, mut index: ResMut<SpatialIndex>| {
+            let mine = query.get(observer.source()).unwrap();
+            let tile = (
+                (mine.pos.x / CELL_SIZE).floor() as i32,
+                (mine.pos.y / CELL_SIZE).floor() as i32,
+            );
+            index.map.entry(tile).or_default().insert(observer.source());
+        },
+    );
+
+    // Since observers run systems you can also define them as standalone functions rather than closures.
+    // This observer runs whenever the `Mine` component is removed from an entity (including despawning it)
+    // and removes it from the spatial index.
+    world.observer(remove_mine);
+
+    // Now we spawn a set of random mines.
+    for _ in 0..1000 {
+        world
+            .spawn(Mine {
+                pos: Vec2::new(
+                    (rand::random::<f32>() - 0.5) * 1200.0,
+                    (rand::random::<f32>() - 0.5) * 600.0,
+                ),
+                size: 4.0 + rand::random::<f32>() * 16.0,
+            })
+            // Observers can also listen to events targeting a specific entity.
+            // This observer listens to `Explode` events targeted at our mine.
+            .observe(
+                |observer: Observer<Explode>, query: Query<&Mine>, mut commands: Commands| {
+                    // If an event is targeting a specific entity you can access it with `.source()`
+                    let source = observer.source();
+                    let Some(mut entity) = commands.get_entity(source) else {
+                        return;
+                    };
```

Comment:
I had an entity validity check in the original dispatch code, but it got removed in a refactor, happy to add it back if we want to ensure that all event sources are alive. Currently two different mines can fire explode events at a specific entity and without this check it will try and create commands for itself after it's been despawned.

Reply:
I see this as more of a symptom of the way we are currently applying commands generated by hooks/observers, if we changed it to only apply new commands after the entire queue is flushed we wouldn't have this issue and the state would generally be more consistent. I think this is a discussion we should have before shipping these features.

Reply:
Fix here: #13249